Q1: 1 – Assuming data is stored in columns from A-D in excel, the new col Net_Ship will be added in col E. Formula: =IF(D-2 < 0, 0, D-2). It prevents invalid entries.<br>
Q1: 2 – select the entire data in excel and insert a pivot table using the Insert tab in ribbon. Drag the Warehouse col in rows and Shipped col in values section of the pivot. Make sure the aggregation is set to Sum.<br>
Q1: 3 – stay on the Shipping col in data and using Ctrl+F, search for -- and in replace send value 0.


In [ ]:
# Q2: 1 – Avg_score_per_group
import pandas as pd
data = {'Trainee': ['T1', 'T2', 'T3', 'T4', 'T5'],
        'Group': ['G1', 'G1', 'G2', 'G2', 'G3'],
        'Python': [ 78, 85, None, 90, 88],
        'Pandas':[80, None, 75,92,85],
        'Excel': [70, 82, 78, None, 90]}
df = pd.DataFrame(data)
df['mean_across_subj'] = df [['Python', 'Pandas', 'Excel']].mean(axis=1, skipna=True)
group_mean = df.groupby('Group') ['mean_across_subj'].mean()
print(df)
print(group_mean)

  Trainee Group  Python  Pandas  Excel  mean_across_subj
0      T1    G1    78.0    80.0   70.0         76.000000
1      T2    G1    85.0     NaN   82.0         83.500000
2      T3    G2     NaN    75.0   78.0         76.500000
3      T4    G2    90.0    92.0    NaN         91.000000
4      T5    G3    88.0    85.0   90.0         87.666667
Group
G1    79.750000
G2    83.750000
G3    87.666667
Name: mean_across_subj, dtype: float64


For trainee T1 -> (78+80+70)/3 = 76.0<br>
For trainee T2 -> (78+80+70)/3 = 76.0<br>
For trainee T3 -> (78+80+70)/3 = 76.0<br>
For trainee T4 -> (78+80+70)/3 = 76.0<br>
For trainee T5 -> (78+80+70)/3 = 76.0

In [ ]:
# Q2: 2 - trainee with maximum total score
df = df.fillna(0,axis=0)
df['total_scores'] = df['Python']+df['Pandas']+df['Excel']
print(df)
print(f"Max total score {df['total_scores'].max()}")

  Trainee Group  Python  Pandas  Excel  mean_across_subj  total_scores
0      T1    G1    78.0    80.0   70.0         76.000000         228.0
1      T2    G1    85.0     0.0   82.0         83.500000         167.0
2      T3    G2     0.0    75.0   78.0         76.500000         153.0
3      T4    G2    90.0    92.0    0.0         91.000000         182.0
4      T5    G3    88.0    85.0   90.0         87.666667         263.0
Max total score 263.0


In [9]:
#Q2: 3 – filling missing values with mean.

df['Python']=df['Python'].fillna(df['Python'].mean())
df['Pandas']=df['Pandas'].fillna(df['Pandas'].mean())
df['Excel']=df['Excel'].fillna(df['Excel'].mean())
df

Trainee Group  Python  Pandas  Excel  mean_across_subj  total_scores
0      T1    G1    78.0    80.0   70.0         76.000000         228.0
1      T2    G1    85.0     0.0   82.0         83.500000         167.0
2      T3    G2     0.0    75.0   78.0         76.500000         153.0
3      T4    G2    90.0    92.0    0.0         91.000000         182.0
4      T5    G3    88.0    85.0   90.0         87.666667         263.0

Q3: 1 – SELECT OrderID, Cust, net_price FROM Orders WHERE net_price = COALESE(Price,0)<br>
Q3: 2 – select Cust, sum(coalese(Price,0)) as total_rev from Orders group by Cust<br>
For cust – C301, total rev = (60+40)<br>
For cust – C301, total rev = (60+40)<br>
For cust – C302, total rev = (0)<br>
For cust – C303, total rev = (120)<br>
Q3: 3 – for missing price values, we can use COALESE(Price,0) which will evaluate to 0, thus ignoring it.

In [11]:
# Q4: 1 - Numpy
import numpy as np

temps = np.array ([55.0, 56.5, 54.0, 57.0, 56.0, 58.0, 57.5])
avg = temps.mean()
print(temps.mean())
print(temps.std())
print(temps.min())

56.285714285714285
1.305404777321219
54.0


In [13]:
print(temps[temps>(avg+0.3)]) 
print(temps[temps>(avg+0.3)].size) 

[57.  58.  57.5]
3


Mean = (55.0 + 56.5 + 54.0 + 57.0 + 56.0 + 58.0 + 57.5)/7 = 56.28<br>
Std = 1.305<br>
Min = minimum value in all values = 54.0<br>

Q4: 2 – print(temps[temps>(avg+0.3)].size) = 3<br>
Mean = 56.28 , mean+0.3 = 56.58<br>
Values satisfying this cond are :  57.0, 58.0, 57.5<br>

In [15]:
# Q4: 3 – Capping values upto an upper limit
temps = temps[temps<101]

Q5: 1 – march -> April <br>
OutletType – mall = 45000-47000 = +2000<br>
OutletType – street = 22000-21000 = -1000<br>
Hence we can see, from mar to apr, the mall outlet type grew by 2000<br>
There was no outlet type change in west, its street in both months.<br>

Q5: 2 – DAX<br>
Tot_sales  = sum(revenue)<br>
Prev_month = calculate(revenue – previous[calendar’month’])<br>
Per_change = Divide((Revenue – previous_month),tot_sales,0)<br>

Q5: 3 – if month is present but a specific date is missing, a random date in that month can be imputed and aggregation over a larger time period can be done like quarter or yearly.<br>

Q6: 1 – for churn = Yes, (3.0+2.9)/2 = 2.95<br>
For churn = No, (4.2+4.7)/2 = 4.45<br>
Q6: 2 – for churn = yes, ticket price (3 -> 5)<br>
for churn = no, ticket price (1 -> 0)<br>
since there is churn happening for increase in ticket price, we can conclude that higher ticket count increases churn.<br>
Q6: 3 – for response time field, ideally the churn should increase with increase in response time.<br>
As for faster response times, the cust should be satisfied.

Q7: 1 – avg amt per userID<br>
For uID – U301 – (30+45)/2 = 37.5<br>
For UID – U302 – avg is 60<br>
For UID – U303 – avg is 20<br>

Q7: 2 – select UserID, rank() over(partition by UserID order by Amt desc) as top_cust from Payments where top_cust=1.

Q7: 3 – for missing payment, we can exclude the record and mentioned about it separately in the document.


In [18]:
# Q8 : 1 - datetime
data = {'emp': ['Ε10', 'E11', 'E12'],
        'start': ['08:00', '08:45', '09:00'],
        'end': ['16:30', '17:15', '17:00']}
df = pd.DataFrame(data)
login = df['start']
pd_login = pd.to_datetime(login, format='mixed')
np_login = pd_login.to_numpy()

logout = df['end']
pd_logout = pd.to_datetime (logout, format='mixed')
np_logout = pd_logout.to_numpy()
df['TotalHrsWorked'] = np_logout - np_login
df

emp  start    end  TotalHrsWorked
0  Ε10  08:00  16:30 0 days 08:30:00
1  E11  08:45  17:15 0 days 08:30:00
2  E12  09:00  17:00 0 days 08:00:00

Q8: 2 – the emps E10 and E11 have equally worked most hours i.e. 8h30mins<br>
Q8: 3 - if isinstance(val, float): hours = int(val);<br>
minutes = round((val - hours) * 100) then format<br>
f"{hours:02d}:{minutes:02d}"<br>
or else, ask to re-enter correctly

Q9: 1 – in excel create a new col with name ‘ctr’ with formula =if(@clicks/@impressions = 0, 0, @clicks/@impressions)<br>
CTR = 240/6000 = 0.04 = 4%<br>
CTR = 160/4000 = 0.04 = 4%<br>
CTR = 250/6100 = 0.04 = 4%<br>

Q9: 2 – A line chart would be insightful as dates are involved. Keep dates on x axis and AdX on y axis.<br>
Yes aggregation would be necessary as there can be data for same dates<br>

Q9: 3 – Zero in impression would cause incorrect computations hence we can impute with (Date+Ad) combination average there.

Q10: 1 – profit = monthlySales - opCost<br>
City P  profit = 320000-260000 = 60000<br>
CityQ profit = 210000- 140000 = 70000<br>
CityR profit = 140000- 85000 = 55000<br>
Since city P & Q have higher profits they can be picked up.

Q10: 2 – KPI<br>
1)	Profit = monthlySales - opCost<br>
CityP = 320000-260000 = 60000<br>
CityQ profit = 210000- 140000 = 70000<br>
CityR profit = 140000- 85000 = 55000<br>
2)	Profit per visitor = profit/Visitors<br>
CityP = 60000/15000 = 4<br>
CityQ profit = 70000/8000 = 8.75<br>
CityR profit = 55000/4200 = 13.09<br>
3)	Visitors per Million Population = Visitors/ Population<br>
CityP = 15000/4.5 = 3333.33<br>
CityQ profit = 8000/3.0 = 26.667<br>
CityR profit = 4200/1.2 = 3500 <br>

Q10: 3 – For CityP, the Transport factor should be considered as it occurs frequently.
- Check 1 — Incremental Operating Cost Estimate: estimate additional costs due to
congestion (deliveries, staff overtime, parking fees) and recompute profit margin after
those costs.
- Check 2 — Service Level Impact: estimate impact on delivery times / customer
satisfaction using existing data (e.g., average delivery time) — this affects repeat
purchases.
- Check 3 — Pilot / Smaller Format: test a smaller format or pop-up to validate demand
and operational feasibility before full investment.
- Check 4 — Regulatory & Logistics Review: check local restrictions, parking/loading
zones, and supply chain constraints; factor in mitigation costs.
